File: BEmu/BloombergTypes/Element.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="BloombergTypes/Element.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    public abstract class Element
    {
        internal abstract StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false);
        public abstract Name Name { get; }
        public abstract int NumValues { get; }
        public abstract int NumElements { get; }

        public virtual Schema.Datatype Datatype { get { return Schema.Datatype.SEQUENCE; } }
        public virtual SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype); } }

        public override string ToString()
        {
            return this.PrettyPrint(0).ToString();
        }

        protected StringBuilder PrettyPrintHelper(int tabIndent, string value, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            string valueToUse = surroundValueWithQuotes ? string.Format("\"{0}\"", value) : value;

            StringBuilder result = new StringBuilder(string.Format("{0}{1} = {2}{3}", tabs, this.Name, valueToUse, Environment.NewLine));

            return result;
        }

        public virtual bool IsNull
        {
            get { throw new NotImplementedException("Element doesn't support IsNull"); }
        }

        public virtual bool IsArray
        {
            get { throw new NotImplementedException("Element doesn't support IsArray"); }
        }

        public virtual bool IsComplexType
        {
            get { throw new NotImplementedException("Element doesn't support IsComplexType"); }
        }

        #region GET ELEMENT ...
        public Element GetElement(Name name)
        {
            return this.GetElement(name.ToString());
        }

        public virtual Element GetElement(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements by name.  name is case-sensitive.");
        }

        public virtual Element GetElement(int index)
        {
            throw new NotImplementedException("Element doesn't support getting elements by index");
        }

        public virtual IEnumerable<Element> Elements
        {
            get { throw new NotImplementedException("Element doesn't support Elements getter"); }
        }

        public virtual object this[int index]
        {
            get { return this.GetValue(index); }
        }

        public virtual Element this[string name]
        {
            get { return this.GetElement(name); }
        }

        public Element this[Name name]
        {
            get { return this.GetElement(name); }
        }

        public virtual object this[string name, int index]
        {
            get { return this.GetElement(name).GetValue(index); }
        }

        public object this[Name name, int index]
        {
            get { return this.GetElement(name).GetValue(index); }
        }

        public bool HasElement(Name name, bool excludeNullElements = false)
        {
            return this.HasElement(name.ToString(), excludeNullElements);
        }

        public virtual bool HasElement(string name, bool excludeNullElements = false)
        {
            throw new NotImplementedException("Element doesn't support HasElement");
        }

        public char GetElementAsChar(Name name) { return this.GetElementAsChar(name.ToString()); }
        public virtual char GetElementAsChar(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as chars");
        }

        public string GetElementAsString(Name name) { return this.GetElementAsString(name.ToString()); }
        public virtual string GetElementAsString(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as strings");
        }

        public Datetime GetElementAsTime(Name name) { return this.GetElementAsTime(name.ToString()); }
        public virtual Datetime GetElementAsTime(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Times");
        }

        public Datetime GetElementAsDate(Name name) { return this.GetElementAsDate(name.ToString()); }
        public virtual Datetime GetElementAsDate(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Dates");
        }

        public Datetime GetElementAsDatetime(Name name) { return this.GetElementAsDatetime(name.ToString()); }
        public virtual Datetime GetElementAsDatetime(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Datetimes");
        }

        public int GetElementAsInt32(Name name) { return this.GetElementAsInt32(name.ToString()); }
        public virtual int GetElementAsInt32(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Int32");
        }

        public long GetElementAsInt64(Name name) { return this.GetElementAsInt64(name.ToString()); }
        public virtual long GetElementAsInt64(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Int64");
        }

        public double GetElementAsFloat64(Name name) { return this.GetElementAsFloat64(name.ToString()); }
        public virtual double GetElementAsFloat64(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Float64");
        }

        public bool GetElementAsBool(Name name) { return this.GetElementAsBool(name.ToString()); }
        public virtual bool GetElementAsBool(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as bool");
        }

        public float GetElementAsFloat32(Name name) { return this.GetElementAsFloat32(name.ToString()); }
        public virtual float GetElementAsFloat32(string name)
        {
            throw new NotImplementedException("Element doesn't support getting elements as Float32");
        }
        #endregion

        #region GET VALUE ...
        public virtual object GetValue() { return this.GetValue(0); }
        public virtual object GetValue(int index)
        {
            throw new NotImplementedException("Element is not a simple object (index)");
        }

        public virtual Element GetValueAsElement() { return this.GetValueAsElement(0); }
        public virtual Element GetValueAsElement(int index)
        {
            throw new NotImplementedException("Element doesn't support getting values as elements (index)");
        }

        public virtual string GetValueAsString(int i)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsString by index");
        }

        public virtual float GetValueAsFloat32()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsFloat32");
        }

        public virtual double GetValueAsFloat64()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsFloat64");
        }

        public virtual double GetValueAsFloat64(int index)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsFloat64");
        }

        public virtual int GetValueAsInt32()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsInt32");
        }

        public virtual int GetValueAsInt32(int index)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsInt32");
        }

        public virtual long GetValueAsInt64()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsInt64");
        }

        public virtual long GetValueAsInt64(int index)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsInt64");
        }

        public virtual string GetValueAsString()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsString");
        }

        public virtual Datetime GetValueAsDatetime()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsDatetime");
        }

        public virtual Datetime GetValueAsDatetime(int index)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsDatetime");
        }

        public virtual Datetime GetValueAsDate()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsDate");
        }

        public virtual Datetime GetValueAsDate(int index)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsDate");
        }

        public virtual Datetime GetValueAsTime()
        {
            throw new NotImplementedException("Element doesn't support GetValueAsTime");
        }

        public virtual Datetime GetValueAsTime(int index)
        {
            throw new NotImplementedException("Element doesn't support GetValueAsTime");
        }
        #endregion
        
        #region APPEND AND SET
        public virtual Element AppendElement()
        {
            throw new NotImplementedException("Element doesn't support AppendElement");
        }

        public virtual void AppendValue(string value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(string)");
        }

        public virtual void AppendValue(Datetime value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(Datetime)");
        }

        public virtual void AppendValue(float value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(float)");
        }

        public virtual void AppendValue(double value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(double)");
        }

        public virtual void AppendValue(long value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(long)");
        }

        public virtual void AppendValue(int value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(int)");
        }

        public virtual void AppendValue(char value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(char)");
        }

        public virtual void AppendValue(bool value)
        {
            throw new NotImplementedException("Element doesn't support AppendValue(bool)");
        }

        public void SetElement(Name name, object value)
        {
            this.SetElement(name.ToString(), value);
        }

        public virtual void SetElement(string name, object value)
        {
            throw new NotImplementedException("Element doesn't support SetElement(Name name, string value)");
        }
        #endregion

    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/Element.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.Element.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu;

import java.lang.StringBuilder;
import com.bemu.BEmu.Name;
import com.bemu.BEmu.types.IndentType;
import com.bemu.BEmu.Datetime;

public abstract class Element {
	protected abstract StringBuilder prettyPrint(int tabIndent);
	public abstract Name name();
	public abstract int numValues();
	public abstract int numElements();
	
	public String toString()
	{
		return this.prettyPrint(0).toString();
	}

    protected StringBuilder prettyPrintHelper(int tabIndent, String value)
    {
        String tabs = IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder(String.format("%s%s = %s%s", 
        		tabs, this.name(), value, System.getProperty("line.separator")));
        return result;
    }

    public boolean isNull() throws Exception
    {
        throw new Exception("Element doesn't support IsNull");
    }

    public boolean isArray() throws Exception
    {
        throw new Exception("Element doesn't support IsArray");
    }

    public boolean isComplexType() throws Exception
    {
        throw new Exception("Element doesn't support IsComplexType");
    }

    public Object getValue() throws Exception { return this.GetValue(0); }
    public Object GetValue(int index) throws Exception
    {
        throw new Exception("Element is not a simple object (index)");
    }

    public String getElementAsString(Name name) throws Exception { return this.getElementAsString(name.toString()); }
    public String getElementAsString(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as strings");
    }

    public Datetime getElementAsTime(Name name) throws Exception { return this.getElementAsTime(name.toString()); }
    public Datetime getElementAsTime(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Times");
    }

    public Datetime getElementAsDate(Name name) throws Exception { return this.getElementAsDate(name.toString()); }
    public Datetime getElementAsDate(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Dates");
    }

    public Datetime getElementAsDatetime(Name name) throws Exception { return this.getElementAsDatetime(name.toString()); }
    public Datetime getElementAsDatetime(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Datetimes");
    }

    public int getElementAsInt32(Name name) throws Exception { return this.getElementAsInt32(name.toString()); }
    public int getElementAsInt32(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Int32");
    }

    public long getElementAsInt64(Name name) throws Exception { return this.getElementAsInt64(name.toString()); }
    public long getElementAsInt64(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Int64");
    }

    public double getElementAsFloat64(Name name) throws Exception { return this.getElementAsFloat64(name.toString()); }
    public double getElementAsFloat64(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Float64");
    }

    public boolean getElementAsBoolean(Name name) throws Exception { return this.getElementAsBool(name.toString()); }
    public boolean getElementAsBool(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as bool");
    }

    public float getElementAsFloat32(Name name) throws Exception { return this.getElementAsFloat32(name.toString()); }
    public float getElementAsFloat32(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements as Float32");
    }

    public Element getValueAsElement() throws Exception { return this.getValueAsElement(0); }
    public Element getValueAsElement(int index) throws Exception
    {
        throw new Exception("Element doesn't support getting values as elements (index)");
    }

    public String getValueAsString(int i) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsString by index");
    }

    public float getValueAsFloat32() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsFloat32");
    }
    public float getValueAsFloat32(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsFloat32");
    }

    public double getValueAsFloat64() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsFloat64");
    }

    public double getValueAsFloat64(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsFloat64");
    }

    public int getValueAsInt32() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsInt32");
    }

    public int getValueAsInt32(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsInt32");
    }

    public long getValueAsInt64() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsInt64");
    }

    public long getValueAsInt64(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsInt64");
    }

    public String getValueAsString() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsString");
    }

    public Datetime getValueAsDatetime() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsDatetime");
    }

    public Datetime getValueAsDatetime(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsDatetime");
    }

    public Datetime getValueAsDate() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsDate");
    }

    public Datetime getValueAsDate(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsDate");
    }

    public Datetime getValueAsTime() throws Exception
    {
        throw new Exception("Element doesn't support getValueAsTime");
    }

    public Datetime getValueAsTime(int index) throws Exception
    {
        throw new Exception("Element doesn't support getValueAsTime");
    }
    

    public Element getElement(Name name) throws Exception
    {
        return this.getElement(name.toString());
    }

    public Element getElement(String name) throws Exception
    {
        throw new Exception("Element doesn't support getting elements by name");
    }

    public Element getElement(int index) throws Exception
    {
        throw new Exception("Element doesn't support getting elements by index");
    }

    public boolean hasElement(Name name) throws Exception
    {
        return this.hasElement(name.toString(), false);
    }

    public boolean hasElement(String name) throws Exception
    {
        throw new Exception("Element doesn't support HasElement");
    }

    public boolean hasElement(Name name, boolean excludeNullElements) throws Exception
    {
        return this.hasElement(name.toString(), excludeNullElements);
    }

    public boolean hasElement(String name, boolean excludeNullElements) throws Exception
    {
        throw new Exception("Element doesn't support HasElement");
    }

    public Element appendElement() throws Exception
    {
        throw new Exception("Element doesn't support AppendElement");
    }

    public void setElement(Name name, Object value) throws Exception
    {
        this.setElement(name.toString(), value);
    }

    public void setElement(String name, Object value) throws Exception
    {
        throw new Exception("Element doesn't support SetElement(Name name, string value)");
    }
}
